# Data Analysis (1): Pandas 1

pandas is an open source library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language. 

pandas is widely used across the scientific and data-analysis communities.

There are many many online resources providing how-to-guides, videos, example-notebooks, etc, etc. 

Examples include: 
 - [https://pandas.pydata.org/docs/index.html](https://pandas.pydata.org/docs/index.html)
   - [https://pandas.pydata.org/docs/user_guide/index.html](https://pandas.pydata.org/docs/user_guide/index.html)
   - [https://pandas.pydata.org/docs/getting_started/intro_tutorials/index.html](https://pandas.pydata.org/docs/getting_started/intro_tutorials/index.html)
   - [https://pandas.pydata.org/docs/getting_started/tutorials.html](https://pandas.pydata.org/docs/getting_started/tutorials.html)
   
 - [https://github.com/guipsamora/pandas_exercises](https://github.com/guipsamora/pandas_exercises)
 
## Contents 
In this `data_analysis_1` notebook we focus on :
1. Basic Data Import using `pandas`
2. Basic Data Exploration using `pandas`

In [5]:
# Imports ...
import numpy as np
import pandas as pd

# 1. Basic Data Import & Properties

In the notebook `fundamentals_14.ipynb` we briefly illustrated how to use the pandas `read_csv` and `to_csv` functions. 

We will now use the `read_csv` function to read a file, but this time we will directly read a file from an *online* location ([https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv](https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv)), rather than from a *local* file.
 - This is a file that is published for use within the tutorials at [pandas.pydata.org](https://pandas.pydata.org/docs/getting_started/intro_tutorials/02_read_write.html)
 - The data consists of the following data columns:
   - PassengerId: Id of every passenger.
   - Survived: Indication whether passenger survived. 0 for yes and 1 for no.
   - Pclass: One out of the 3 ticket classes: Class 1, Class 2 and Class 3.
   - Name: Name of passenger.
   - Sex: Gender of passenger.
   - Age: Age of passenger in years.
   - SibSp: Number of siblings or spouses aboard.
   - Parch: Number of parents or children aboard.
   - Ticket: Ticket number of passenger.
   - Fare: Indicating the fare.
   - Cabin: Cabin number of passenger.
   - Embarked: Port of embarkation.

In [6]:
titanic = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv")

The above file-read should have silently populated the `titanic` variable. 

We can now use some of the essential pandas tools to understand the data that has been imported. 

# 1.1 Dataframe Properties

 - The imported variable, `titanic` is a pandas `Dataframe` 

In [7]:
# Illustrate that *titanic* is a *dataframe*
print( " type = ... ", type(titanic) )

 type = ...  <class 'pandas.core.frame.DataFrame'>


 - The dataframe function `inof()` will give us a handy summary of `titanic`, it's type, the fields it contains, the number of data entries, etc:

In [8]:
"""
Illustrate the use of the *info()* function

Lots of useful info returned, including ... 
 - It is indeed a DataFrame.
 - There are 891 entries, i.e. 891 rows.
 - Each row has a row label/index, with values ranging from 0 to 890.
 - The table has 12 columns. 
 - Most columns have a value for each of the rows (all 891 values are non-null). 
 - Some columns have missing values (less than 891 non-null values).
 - The columns Name, Sex, Cabin and Embarked consists of textual data (strings, aka object). 
 - The other columns are numerical data with some of them whole numbers (aka integer) and others are real numbers (aka float).
 - The kind of data (characters, integers,…) in the different columns are summarized by listing the dtypes (data-types).
 - The approximate amount of RAM used to hold the DataFrame is provided as well.
"""
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


# 1.2 Head & Tail 
 - `head()` and `tail()` illustrate the first and last few lines of the data that has been imported
 - By default they show 5 lines, but you can change that by supplying a numerical argument

In [9]:
# Illustrate calling the *head* function with *no* argument
# - 5 data rows are returned (plus the header information)
titanic.head() 

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [10]:
# Illustrate calling the *tail* function with a numerical value of 3 for the (optional) argument
# - 3 data rows are returned (plus the header information)
titanic.tail(3) 

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


# 1.3 Describe 
 - The `describe()` function provides a numerical summary of any *numerical* fields in the table

In [11]:
# Calling the *describe()* function to illustrate the summary that is provide for the 7 numerical fields in the dataframe 
titanic.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


# Exercise Pandas: 1(a) 

Reading a file with a header...

1. Read the online csv file [https://people.sc.fsu.edu/~jburkardt/data/csv/airtravel.csv](https://people.sc.fsu.edu/~jburkardt/data/csv/airtravel.csv) into a dataframe 
2. Use the `info()` function to understand the names of the columns

# Exercise Pandas: 1(b) 

Read a csv file without a header...

1. Read the online csv file [https://people.sc.fsu.edu/~jburkardt/data/csv/addresses.csv](https://people.sc.fsu.edu/~jburkardt/data/csv/addresses.csv) into a dataframe 
2. Use the `head()` function to look at the data
 - What does it look like? 
 - Did you tell the `read_csv` function what the column *names* are? 
 - If not, the `read_csv` function will have incorrectly assumed the names are in the first row. 
3. Lookup `names` in the [online documentation](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) for the `read_csv` function
 - Now add `names` list when you call `read_csv`, and re-import the data 

# Exercise: Data Analysis 1(A) 

Sometimes data might be supplied in a *fixed-width* format. 

Here the data are *not* separated/delimited by commas, but instead are confined to columns of specified widths.

The example provided in [pandas_1_ex1c_fwf.txt](pandas_1_ex1c_fwf.txt) looks like ...

|   name  |  age  | distance |
|---------|-------|----------|
|   bob   |   1.2 |  10.5    |
|cleo     |  3.3  |   5.91   |
| dirk    | 11.29 | 2.31     |


1. Look at the online documentation to understand how to use the `read_fwf` 
function. 
 - [https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.read_fwf.html](https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.read_fwf.html)
 - [online documentation for the read_fwf function](https://pandas.pydata.org/docs/reference/api/pandas.read_fwf.html)
2. Use the `read_fwf` function to read the example file []() into a dataframe. It might be useful to understand how to use the following options:
 - skiprows
 - header
 - widths
3. Use the `info()` function to understand the properties of the dataframe. 
4. Use the `head()` function to illustrate some sample rows of the data. 

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   name      3 non-null      object 
 1   age       3 non-null      float64
 2   distance  3 non-null      float64
dtypes: float64(2), object(1)
memory usage: 200.0+ bytes


In [17]:
df.head()

,name,age,distance
0,bob,41.20,10.50
1,cleo,53.30,5.91
2,dirk,21.29,2.31


# 2. Basic Data Exploration : Selecting Data

## 2.1. Selecting Data


Let's use the `titanic` dataframe to learn how to do some basic data exploration using built-in methods from the `pandas` package

In [18]:
titanic = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv")

## 2.1.1: Selecting by Column Label/Name

We saw above (from using the `info` and/or `head` functions) that one of the fields in the `titanic` dataframe is called "Age".

We can use the field-labels to select just the data in that one field.
 - In the example below we select the field just as we do in a dictionar.

The data that is returned is a list-like object which has a type `series`

In [22]:
# Select the field named "Age"
s = titanic["Age"]

# Print out the information for the returned object
s.describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

You can also select the data in a column as follows:
`s = titanic.Age`

I.e. the dataframe has a named "attribute", `Age`, that you can select

In [21]:
s = titanic.Age
s.describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

In [41]:
# The `head` function also works on series (as well as the full dataframe)
s.head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

## 2.1.2: Selecting Rows: Slice Notation

We can use the slice notation from "fundamentals_4.ipynb" to select a subset of the rows from a dataframe 

In [5]:
# Select rows with indicees 3-to-6 (inclusive)
rs = titanic[3:7]

# The returned object is a dataframe 
rs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 3 to 6
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  4 non-null      int64  
 1   Survived     4 non-null      int64  
 2   Pclass       4 non-null      int64  
 3   Name         4 non-null      object 
 4   Sex          4 non-null      object 
 5   Age          3 non-null      float64
 6   SibSp        4 non-null      int64  
 7   Parch        4 non-null      int64  
 8   Ticket       4 non-null      object 
 9   Fare         4 non-null      float64
 10  Cabin        2 non-null      object 
 11  Embarked     4 non-null      object 
dtypes: float64(2), int64(5), object(5)
memory usage: 516.0+ bytes


In [6]:
# The returned object 
rs

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S


## 2.1.3 Selecting Rows by Condition

pandas makes it very easy to select subsets of the rows in a dataframe according to the values in certain fields. 

As an example, in the cells below we will use the above `titanic` dataframe, and we will select all passengers over the age of 20. 

In [71]:
# Create the dataframe 
titanic = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv")
# Create a *Series* of booleans (True/False) that labels whether or not the value in the `Age` field is > 20 
s = titanic['Age'] > 20
s

0       True
1       True
2       True
3       True
4       True
       ...  
886     True
887    False
888    False
889     True
890     True
Name: Age, Length: 891, dtype: bool

In [74]:
# Use the Series, `s`, to select a subset of the rows in the dataframe 
above_twenty = titanic[s]
print(f'len(titanic) = {len(titanic)}')
print(f'len(above_twenty) = {len(above_twenty)}')
above_twenty


len(titanic) = 891
len(above_twenty) = 535


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
884,885,0,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


# You can combine the above steps that create a series and then select the rows into a *single step*

In [75]:
# You can combine the creation of the series and the selection of the rows into a single step
above_twenty = titanic[titanic['Age'] > 20]
above_twenty

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
884,885,0,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


# Exercise: Data Analysis 1(B) 

Import the ["https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv"]("https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv") file into a dataframe 

1. How many rows are in the dataframe ? 
2. Create a new dataframe containing only the rows for passangers whose fare was more than \$20 *and* less then \$30
 - N.B. To combine *2* series you can use the notation `series_c = series_a & series_b`
   - Use this approach to combine one series for `>20` with a second series for `<30`
3. How many rows are in the new dataframe?
4. Print out rows 100 - 109 (inclusive) for the new dataframe 

## 2.2. Basic Data Exploration : Sorting Data

The `DataFrame.sort_values()` method is used to sort a DataFrame by its column or row values. 

In [44]:
titanic.head() 

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [48]:
sorted_titanic = titanic.sort_values(by="Age")
sorted_titanic.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
803,804,1,3,"Thomas, Master. Assad Alexander",male,0.42,0,1,2625,8.5167,NaN,C
755,756,1,2,"Hamalainen, Master. Viljo",male,0.67,1,1,250649,14.5000,NaN,S
644,645,1,3,"Baclini, Miss. Eugenie",female,0.75,2,1,2666,19.2583,NaN,C
469,470,1,3,"Baclini, Miss. Helene Barbara",female,0.75,2,1,2666,19.2583,NaN,C
78,79,1,2,"Caldwell, Master. Alden Gates",male,0.83,0,2,248738,29.0000,NaN,S


Note from the table above, that after using the `sort_values` command, the youngest passengers now appear at the top of the table. 

The `by` parameter can take a list of column names.
 - When a list of columns is supplied, the table is sorted first using the first field in the list, and then if any items in the first field are the same, the second field in the list is used as the basis for sorting. 


In [51]:
sorted_titanic2 = titanic.sort_values(by=["Age","PassengerId"])
sorted_titanic2.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
803,804,1,3,"Thomas, Master. Assad Alexander",male,0.42,0,1,2625,8.5167,NaN,C
755,756,1,2,"Hamalainen, Master. Viljo",male,0.67,1,1,250649,14.5000,NaN,S
469,470,1,3,"Baclini, Miss. Helene Barbara",female,0.75,2,1,2666,19.2583,NaN,C
644,645,1,3,"Baclini, Miss. Eugenie",female,0.75,2,1,2666,19.2583,NaN,C
78,79,1,2,"Caldwell, Master. Alden Gates",male,0.83,0,2,248738,29.0000,NaN,S


# Exercise: Data Analysis 1(C) 

A dictionary, `stellar_dict`, is provided below. 
Use `stellar_dict` to create a dataframe. 
Sort the dataframe primarily by "Age", secondarily by "Mass" and then finally by "Distance"

In [8]:
stellar_dict = {
    "Age":     [3.0, 4.0, 3.0, 5.0, 4.0, 5.0, 3.5, 3.5, 4.5, 4.0 ],
    "Mass":    [0.3, 0.4, 0.5, 0.4, 0.3, 0.3, 0.5, 0.4, 0.3, 0.4],
    "Distance":[5.5, 7.6, 5.1, 9.9, 7.7, 4.9, 5.5, 8.1, 6.2, 6.0],
    "Label":   ['A', 'B', 'C', 'D', 'E', 'F',  'G','H', 'I', 'J'],
}